File: 22/day1/code.js
const fs = require("fs");
const readline = require("readline");

let top_3_counts = [0, 0, 0];
let running_total = 0;

/***
 * This code uses streams to do the work, reading a line at a time rather than everything at once. Better memory management and less overall looping.
 */
const reader = readline.createInterface({
  input: fs.createReadStream("./input.txt"),
});

reader.on("line", (line) => {
  if (!line.length) {
    top_3_counts.push(running_total);
    top_3_counts.sort();
    top_3_counts.shift();
    running_total = 0;
  } else {
    running_total += parseInt(line);
  }
});

reader.once("close", () => {
  console.log("Maximum:", top_3_counts[top_3_counts.length - 1]);
  console.log(
    "Top three:",
    top_3_counts[0] + top_3_counts[1] + top_3_counts[2]
  );
});

/***
 * This code is easier to understand, but will use more memory and will need to loop over the entries twice, once for file load and once to process the content.
 */
const lines = fs.readFileSync("./input.txt").toString().split("\n");
let counts = [];
running_total = 0;

for (const line of lines) {
  if (!line.length) {
    counts.push(running_total);
    running_total = 0;
  } else {
    running_total += parseInt(line);
  }
}
counts.sort().reverse();
console.log("Maximum:", counts[0]);
console.log("Top three:", counts[0] + counts[1] + counts[2]);

File: 22/day1/code.rs
use std::fs::File;
use std::io::{self, prelude::*, BufReader};

fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = BufReader::new(file);

    let mut counts = Vec::new();
    let mut running_total: u32 = 0;

    for line in reader.lines() {
        match line {
            Ok(value) => {
                match value.len() {
                    0 => {
                        counts.push(running_total);
                        running_total = 0;
                    },
                    _ => running_total += value.parse::<u32>().unwrap(),
                }
            },
            Err(_) => panic!("Error reading file!")
        }
    }

    counts.sort_by(|a, b| b.cmp(a));

    println!("Maximum: {}", counts[0]);
    println!("Top three: {}", counts[0] + counts[1] + counts[2]);

    Ok(())
}